lib/services/database.go

```diff
@@ -1402,16 +1513,35 @@ func GetMySQLEngineVersion(labels map[string]string) string {
 	return version
 }
 
+// IsAzureFlexServer returns true if the database engine label matches the Azure PostgreSQL or MySQL Flex server engine name.
+// Matching engines are "Microsoft.DBforMySQL/flexibleServers" or "Microsoft.DBforPostgreSQL/flexibleServers".
+func IsAzureFlexServer(db types.Database) bool {
+	engine, ok := db.GetMetadata().Labels[labelEngine]
+	return ok && (engine == AzureEngineMySQLFlex || engine == AzureEnginePostgresFlex)
+}
+
+// MakeAzureDatabaseLoginUsername returns a user name appropriate for Azure database logins.
+// Azure requires database login to be <user>@<server-name>,
+// for example: alice@mysql-server-name.
+// Flexible server is an exception to this format and returns the provided username unmodified.
+func MakeAzureDatabaseLoginUsername(db types.Database, user string) string {
+	// https://learn.microsoft.com/en-us/azure/mysql/flexible-server/how-to-azure-ad
+	if IsAzureFlexServer(db) {
+		return user
+	}
+	return fmt.Sprintf("%v@%v", user, db.GetAzure().Name)
```

Comment:
Ah, good catch! How about instead of adding a flag for flex server to `database[].azure.IsFlexiServer`, we enforce that an Azure endpoint specifies a resource ID as well. We check if the endpoint is an azure db here:
https://github.com/gravitational/teleport/blob/f942a4ebc8d9a1e3158b05061b5c2174ccf4e2b1/api/types/database.go#L606-L614

I think we should check `d.Spec.Azure.ResourceID != ""`. If we enforce that requirement, then we can parse the resource ID to extract the db provider, e.g. `Microsoft.DBforMySQL/flexibleServers`, which is where the engine label comes from.

Reply:
This change would break existing configs for static registering single-server though. Let's discuss in 1x1 tomorrow.

Reply:
@smallinsky and I discussed in 1x1. Making `ResourceID` required would break all existing static Azure database configs in v12 release. Flexi server connection via tsh does not work in v11, due to the difference in username format, so there should not be any existing (usable) static flexi server databases in configs. Therefore, it would be better to just add the `IsFlexibleServer` boolean to `azure:` section of static db config.

When single-server goes completely end-of-life, this boolean option will become obsolete and we will just always use the flexi server username format for connections. That will be in 2024: https://learn.microsoft.com/en-us/azure/mysql/single-server/whats-happening-to-mysql-single-server

> Hello! We have news to share - Azure Database for MySQL - Single Server is on the retirement path and Azure Database for MySQL - Single Server is scheduled for retirement by September 16, 2024.
> 
> As part of this retirement, we will no longer support creating new Single Server instances from the Azure portal beginning January 16, 2023. If you still need to create Single Server instances to meet business continuity needs, you can leverage [Azure CLI](https://learn.microsoft.com/en-us/azure/mysql/single-server/quickstart-create-mysql-server-database-using-azure-cli). Additionally, you can still use your Terraform template to create single server instances. You will still be able to create read replicas for your existing single server instance from the Replication blade and this will continue to be supported till the sunset date of September 16, 2024.

Reply:
@smallinsky @greedy52 On this topic, I thought more about Azure permissions bootstrapping. I think I will draft an RFD for that this week. One of the objections to having resourceid required and adding Azure metadata fetching is that the service principal will require permissions that static Azure databases did not require previously. If we add bootstrapping, then adding the needed permissions to the service principal will be easy for users to do. What are your thoughts on that?

Reply:
agreed having the flag for azure block is the simplest solution for now, and it's easier for users than using the resource id. I don't think we can make this new flag required for the same backwards compatibility reason either, but maybe we can mark it like "required for flex servers" in the reference doc.

Reply:
Yeah that's what I was thinking. Since it's only required for flex servers, and users should not have any flex servers in their configs on prior teleport versions (since it wouldn't work), then it should be fine to add it as a required field for flex config

Reply:
added the flag and changed the base of this PR to point at the protobuf PR for the bool flag: https://github.com/gravitational/teleport/pull/19970